### PR TITLE
[3007.x] Fix Salt-SSH loader attribute access, document that it's still not recommended

### PR DESCRIPTION
**Pull request by lkubb**
### What does this PR do?
* Ensures the `FunctionWrapper` used in Salt-SSH templates behaves more like the regular LazyLoader
* Adjusts documentation that `salt['foo.bar']()` is to be preferred over `salt.foo.bar()` since the latter can introduce edge cases

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/41794
Fixes: https://github.com/saltstack/salt/issues/66376
Fixes: https://github.com/saltstack/salt/issues/66600

### Previous Behavior
Several issues when using the `salt.foo.bar` syntax (see linked issues)

### New Behavior
Works as expected. There might be some unknown edge cases though, so still prefer `salt['foo.bar']`.

### Merge requirements satisfied?
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

File: doc/ref/states/requisites.rst [hunk 1]
```diff
@@ -563,7 +563,7 @@ The ``onfail`` requisite is applied in the same way as ``require`` and ``watch``
     notify-build_failure:
       hipchat.send_message:
         - room_id: 123456
-        - message: "Building website fail on {{ salt.grains.get('id') }}"
+        - message: "Building website fail on {{ salt['grains.get']('id') }}"
```
Comment (OrangeDog) [inline doc/ref/states/requisites.rst:566]: This one should probably be `{{ grains["id"] }}`.

File: doc/ref/states/requisites.rst [hunk 2]
```diff
@@ -723,7 +723,7 @@ be installed. Thus allowing for a requisite to be defined "after the fact".
 
 .. code-block:: sls
 
-    {% for cfile in salt.pillar.get('nginx:config_files') %}
+    {% for cfile in salt['pillar.get']('nginx:config_files') %}
```
Comment (OrangeDog) [inline doc/ref/states/requisites.rst:726]: `black` rules in favour of double quotes. Documentation should ideally be consistent with that.

```py
salt["pillar.get"]("nginx:config_files")
```

There are other examples in this file and PR.

Comment (OrangeDog) [reply doc/ref/states/requisites.rst:726]: Oh, well nevermind then.

**Review by Akm0d — APPROVED** [review]

**Review by Akm0d — APPROVED** [review]

File: salt/states/file.py [hunk 3]
```diff
@@ -8786,7 +8786,7 @@ def decode(
             - name: /tmp/new_file
             - encoding_type: base64
             - encoded_data: |
-                {{ salt.pillar.get('path:to:data') | indent(8) }}
```
Comment (dmurphy18) [inline salt/states/file.py:8789]: see above for disagree

**Review by dmurphy18 — COMMENTED** [review]
issues with best practices being reverted when accessing essentially dictionaries

File: doc/ref/states/requisites.rst [hunk 4]
```diff
@@ -563,7 +563,7 @@ The ``onfail`` requisite is applied in the same way as ``require`` and ``watch``
     notify-build_failure:
       hipchat.send_message:
         - room_id: 123456
-        - message: "Building website fail on {{ salt.grains.get('id') }}"
```
[the remark this answers is not in the document]
Comment (OrangeDog) [reply doc/ref/states/requisites.rst:566]: Disagree. Unless you are deliberately handling the `None` you _want_ it to raise an error rather than continue silently and then either fail somewhere else or do something bad. Therefore `[]` is preferred in the general case.

File: doc/topics/best_practices.rst [hunk 5]
```diff
@@ -21,6 +21,13 @@ General rules
 4. Store sensitive data in pillar.
 5. Don't use grains for matching in your pillar top file for any sensitive
    pillars.
+6. When accessing modules from within a template, use the mapping
+   key syntax instead of the attribute one to avoid edge cases. Example:
+
+   .. code-block:: jinja
+
+       {%- set do_this = salt['pillar.get']('foo:bar') %}
```
[the remark this answers is not in the document]
Comment (lkubb) [reply doc/topics/best_practices.rst:29]: Is this about the missing default value? Otherwise not sure how the `.get` comment above applies from here on. This example is not about accessing dict keys, but how to use Salt execution modules in Jinja templates in general. `salt.pillar.get` should be avoided, the same with `salt.cmd.run` etc. `salt["pillar.get"]` or `salt["cmd.run"]` don't have weird edge cases and should thus be preferred.

I'll answer the other disagrees here as well:

`salt["pillar.get"]` additionally traverses into subdicts, so the `pillar.get` version would look like this:

`{{ pillar.get("myapp", {}).get("version", "master") }}`

or

`{{ pillar | traverse("myapp:version", "master") }}`

Changing from `salt.pillar.get` to `salt["pillar.get"]` only changes the way the execution module is accessed, not other semantics.

File: doc/ref/states/requisites.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/ref/states/requisites.rst:566]: Okay, get your point, better to have it throw an error if the index is missing in whatever['my_missing_key'] than have to deal with side-effects

File: doc/topics/development/conventions/formulas.rst [hunk 6]
```diff
@@ -262,7 +262,7 @@ file. This section contains several suggestions and examples.
     deploy_myapp:
       git.latest:
         - name: git@github.com/myco/myapp.git
-        - version: {{ salt.pillar.get('myapp:version', 'master') }}
```
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/topics/development/conventions/formulas.rst:265]: See response for acceptance above

File: doc/topics/development/conventions/formulas.rst [hunk 7]
```diff
@@ -690,7 +690,7 @@ Macros are useful for creating reusable, parameterized states. For example:
         - groups: {{ groups | json() }}
     {% endmacro %}
 
-    {% for user_info in salt.pillar.get('my_users', []) %}
```
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/topics/development/conventions/formulas.rst:693]: See response for acceptance above

File: doc/topics/development/conventions/formulas.rst [hunk 8]
```diff
@@ -1033,11 +1033,11 @@ example:
 
     {# Extract the relevant subset for the app configured on the current
        machine (configured via a grain in this example). #}
-    {% app = app_defaults.get(salt.grains.get('role')) %}
+    {% app = app_defaults.get(salt['grains.get']('role')) %}
 
     {# Allow values from Pillar to (optionally) update values from the lookup
        table. #}
-    {% do app_defaults.update(salt.pillar.get('myapp', {})) %}
```
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/topics/development/conventions/formulas.rst:1040]: See response for acceptance above

File: doc/topics/tutorials/jinja_to_execution_module.rst [hunk 9]
```diff
@@ -52,7 +52,7 @@ Unfortunately, it can lead to code that looks like the following.
     {%   do storage.update({'server_ip': servers_list[server_index]}) %}
     {% endif %}
 
-    {% for network, _ in salt.pillar.get('inventory:networks', {}) | dictsort %}
```
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/topics/tutorials/jinja_to_execution_module.rst:55]: See response for acceptance above

File: salt/states/file.py (hunk 3 above)
Comment (dmurphy18) [reply salt/states/file.py:8789]: See response for acceptance above

**Review by dmurphy18 — APPROVED** [review]

File: salt/states/file.py (hunk 3 above)
Comment (dmurphy18) [reply salt/states/file.py:8789]: See response for acceptance above

File: doc/topics/tutorials/jinja_to_execution_module.rst [hunk 10]
```diff
@@ -88,7 +88,7 @@ Let's move that to an execution module.
 
     {% do storage.update({'server_ip': salt['storage.ip']()}) %}
 
-    {% for network, _ in salt.pillar.get('inventory:networks', {}) | dictsort %}
```
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/topics/tutorials/jinja_to_execution_module.rst:91]: See response for acceptance above

File: doc/topics/reactor/index.rst [hunk 11]
```diff
@@ -394,8 +394,8 @@ For example:
 .. code-block:: jinja
 
     # /srv/salt/orchestrate/do_complex_thing.sls
-    {% set tag = salt.pillar.get('event_tag') %}
```
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/topics/reactor/index.rst:397]: See response for acceptance above

File: doc/topics/development/conventions/formulas.rst [hunk 12]
```diff
@@ -478,7 +478,7 @@ Below is a simple example of a readable loop:
 
 .. code-block:: jinja
 
-    {% for user in salt.pillar.get('list_of_users', []) %}
```
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/topics/development/conventions/formulas.rst:481]: See response for acceptance above

File: doc/topics/development/conventions/formulas.rst [hunk 13]
```diff
@@ -920,7 +920,7 @@ Pillar can also be used.
 .. code-block:: jinja
 
     {% set lookup_table = {...} %}
-    {% do lookup_table.update(salt.pillar.get('my:custom:data')) %}
```
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/topics/development/conventions/formulas.rst:923]: See response for acceptance above

File: doc/topics/development/conventions/formulas.rst [hunk 14]
```diff
@@ -994,7 +994,7 @@ XML.)
 .. code-block:: jinja
 
     {% import_yaml 'tomcat/defaults.yaml' as server_xml_defaults %}
-    {% set server_xml_final_values = salt.pillar.get(
```
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/topics/development/conventions/formulas.rst:997]: See response for acceptance above

File: doc/topics/development/conventions/formulas.rst [hunk 15]
```diff
@@ -1033,11 +1033,11 @@ example:
 
     {# Extract the relevant subset for the app configured on the current
        machine (configured via a grain in this example). #}
-    {% app = app_defaults.get(salt.grains.get('role')) %}
```
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/topics/development/conventions/formulas.rst:1036]: See response for acceptance above

File: doc/topics/development/conventions/formulas.rst [hunk 16]
```diff
@@ -708,7 +708,7 @@ example, the following macro could be used to write a php.ini config file:
         - source: salt://php.ini.tmpl
         - template: jinja
         - context:
-            php_ini_settings: {{ salt.pillar.get('php_ini', {}) | json() }}
```
[the remark this answers is not in the document]
Comment (dmurphy18) [reply doc/topics/development/conventions/formulas.rst:711]: See response for acceptance above

**Review by twangboy — APPROVED** [review]

**Review by Akm0d — APPROVED** [review]
